### PR TITLE
Queue::run() - Add API for running several tasks

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

Adds a utility for running queued tasks using the newer (higher-level) queuing features. For example:

```bash
cv api4 Queue.run queue=my-stuff maxDuration=120 maxRequests=100
```

(*This was originally written with name `Queue.runLoop`. The discussion goes into wordsmithing and lands on `Queue.run`. The description is updated to reflect the final name, but the discussion may be easier to read if you know the original name.*)

Before
----------------------------------------

* You can use `Civi\Api4\Queue::runItems()` to claim and execute one unit-of-work (i.e. task or batch of items).

After
----------------------------------------

* You can use `Civi\Api4\Queue::runItems()` to claim and execute one unit-of-work (i.e. task or batch of items).
* You can use `Civi\Api4\Queue::run()` to loop through a series of many units-of-work (i.e. several tasks or several batches).

Technical Details
----------------------------------------

As [discussed in MM](https://chat.civicrm.org/civicrm/pl/pr98jpu68fb5unb8z1xnfp7iwy), it's useful to provide more parity in the Queue APIv4 world. This addresses the first of two major gaps in the functionality of Queue APIv4 world. (The other being `runAllViaWeb()`.) (ping @eileenmcnaughton)

`Civi\Api4\Queue::run()` is analogous to the older `CRM_Queue_Runner::runAll()`. It is also analogous to the `civicrm_api_job_runspecificqueue` template suggested by dev-docs. However, this variant has more features:

| | `CRM_Queue_Runner::runAll()` | `civicrm_api_job_runspecificqueue` | `Civi\Api4\Queue::run()` |
| -- | -- | --| -- |
| Loop through items in a queue | Yes | Yes | Yes |
| Support all config flags (eg `batch_limit`) | No | No | Yes |
| Support `hook_queueRun_XXX` (*batching*) | No |  No |Yes |
| Support `hook_queueTaskError` | No |  No |Yes |
| Support `civi.queue.check` | No |  No |Yes |
| Support time-limit (*soft*) | No | Yes | Yes
| Support iteration-limit | No | No | Yes

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@totten elsewhere we use `runAll` - how does `runLoop` differ conceptually - does it only run some?

**Comment by eileenmcnaughton** [comment]
@totten why don't we call it `runAll`? ie new `runAll` is old `runAll`++ from what you say as opposed to something conceptually different - which is what I am thinking `runLoop` is

File: Civi/Api4/Action/Queue/RunLoop.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        break;
+      }
+
+      foreach ($batchResults as $batchResult) {
+        if ($batchResult['outcome'] === 'ok') {
+          $successes++;
+        }
+        else {
+          $errors++;
+          // Should we stop? No, we're just reporting stats.
```
Comment (eileenmcnaughton) [inline Civi/Api4/Action/Queue/RunLoop.php:105]: So - reading this - we would continue to process items from a DIFFERENT queue but not the same one wouldn't we?

Comment (totten) [reply Civi/Api4/Action/Queue/RunLoop.php:105]: @eileenmcnaughton Good q. It should break down like this

* In the call `cv api4 Queue.runLoop queue=my-stuff `, it's scoped to one specific queue. The `queue` parameter is required. 
* If that queue has `error=>abort` policy and then becomes `status=>aborted`, then `runLoop` stops.
* You could have two different queues running two copies of `Queue.runLoop` . Then yes -- the other queue continues executing.

(*Note to self: It should probably return a better message. All the other exit conditions have a descriptive message to facilitate debugging...*)

Comment (eileenmcnaughton) [reply Civi/Api4/Action/Queue/RunLoop.php:106]: so @totten `runLoop` is basically `runQueue` & `runAll` is `runAll` queues ?

**Comment by totten** [comment]
> (@eileenmcnaughton) why don't we call it `runAll`? ie new `runAll` is old `runAll`++ from what you say as opposed to something conceptually different - which is what I am thinking `runLoop` is

Yeah, `CRM_Queue_Runner::runAll()` and `Civi\Api4\Queue::runLoop()` are the same basic idea. I would be OK with calling it `runAll()`, and I've posted another PR (#27709) to run Jenkins with that name.

But let me give a pitch for why `runLoop` is better than `runAll`:

* `Civi\Api4\Queue::runLoop()` is also the same basic idea as the example `civicrm_api_job_runspecificqueue` (dev-docs)... which is a bit different from `CRM_Queue_Runner::runAll()`.
* I vaguely recall some previous feedback like: "`CRM_Queue_Runner::runAll` doesn't really run _everything_. It only runs the _things that are there right now_." (The word "all" was chosen when the upgrader was the only use-case, and that was a fixed-size scope of work. But there are other things -- like time-based release or backgrounded civirules -- which are open-scope.)
* `runLoop()` has parameters `maxRequests` and `maxDuration` which act as additional limits on how many tasks it runs. (*In current draft, defaults are 10 requests or 120 seconds, whichever comes sooner.*) But if you have limits, then it's not "all".

I would say the primary goal of the PR is to get functional parity for low-level-runner/high-level-runner. But there is a secondary goal -- add some kind of pressure-relief-valve/hedge against issues in `coworker`. (*Like: What if a sysadmin really wants to throttle queue work-loads in a way that coworker doesn't initially support? Or what if we find some major bug that takes 4 weeks to fix?*) The PRV/hedge doesn't need to be perfect. `runLoop` basically "as good as" the pre-existing `civicrm_api_job_runspecificqueue()` technique. So one *can* setup a similar crontab, eg:

```
0/10  * * * www-data cd /var/www/mysite && cv api4 Queue.runLoop queue=first maxRequests=10
2/10  * * * www-data cd /var/www/mysite && cv api4 Queue.runLoop queue=second maxRequests=20
4/10  * * * www-data cd /var/www/mysite && cv api4 Queue.runLoop queue=third maxRequests=40

## Swap cv / drush / curl to taste
```

On the flip side,  `runLoop` can sound slightly loopy. `runAll` sounds a bit more familiar, and it's still in the ballpark of accurate. So I could go along with that too.

**Comment by eileenmcnaughton** [comment]
"On the flip side, runLoop can sound slightly loopy"

lol

**Comment by eileenmcnaughton** [comment]
@totten OK so trying to get my head round this - there are 2 actions I would want to do. I think of them as `runQueue` (sand `runAll`. Where `runQueue` is a specific queue & `runAll` is whatever queues it finds

I  think you are saying that `runLoop` is actually more similar to `runQueue` conceptually than `runAll`

**Comment by totten** [comment]
@eileenmcnaughton Yes, mapping those phrases, I think `runLoop` <=> `runQueue`, and `coworker` <=>  `runAll`.

The names `runQueue` and `run` did come to mind as well.

*  `cv api4 Queue.runQueue queue=xyz` seemed a bit repetitive.
* For `run`, the niggle against it.... in the context of "several independent queues, each with optional batching support and executable tasks", the bare word "run" can be taken to mean like 6 different things.
    * (But in favor, for `Civi\Api4\Queue::run`, I would plurality of readers to interpret it like "run a main-loop on the given queue".)

For me, these are all roughly at par (all acceptable; just trading fiddliness): `run`, `runLoop`, `runQueue`, `runAll`.

**Comment by eileenmcnaughton** [comment]
@totten yeah - I come down on `run` or `runQueue`. `runQueue` seems to be what it actually does -  but I wasn't clear on that until you confirmed it in your last comment. 

`run` - I think the strength & weakness is the multiple meanings - ie if it takes parameters that dictate whether it runs only one queues or multiple or this or that then `run` is good. But if this really just is about adding an action that will run a specific queue then `run` is probably less clear than `runQueue`

Now that I understand what is being added I agree the functionality is a great addition & test cover is obviously good. I think having this right now on our staging would be really good because I'm testing a big data cleanup on our mailing tables and have queued stuff up & it's not working but then I realised that's because... I've queued up a half-dozen tasks that are re-queuing & maybe in a few hours it will get to my item....

It would be interesting to think about this in terms of the future of `civicrm_job` table.

**Comment by totten** [comment]
> I come down on `run` or `runQueue`

My personal rankings would be like: 

`runLoop > run > runQueue == runAll`

So it sounds like `run` is probably the best compromise. (Unless someone else wants to push the scales toward another name overnight... :eyes:  )

**Comment by colemanw** [comment]
Is it too late to contribute to the wordsmithing? Thinking about my experience in queues (e.g. at airport security) I think maybe `Queue.shuffle()` or `Queue.amble()` or sometimes if you're trying to get to the end of the queue from the other side of the terminal it would be `Queue.powerWalk()`.

**Comment by eileenmcnaughton** [comment]
@colemanw yeah I think amble was probably a good description for what's happening for me now

**Pull request merged**